File: packages/typescript/src/resources/contacts.ts
/**
 * The `contacts` resource: the workspace address book.
 * @module
 */
import { toWire } from '../models.js';
import type { Contact, CreateContactParams, ListParams, Page, RequestOptions, UpdateContactParams } from '../models.js';
import { Resource, seg, wireQuery } from './base.js';

/** Accessed as `client.contacts`. */
export class Contacts extends Resource {
  list(params: ListParams = {}, options?: RequestOptions): Promise<Page<Contact>> {
    return this.page({ method: 'GET', path: '/v1/contacts', query: wireQuery(params) }, options);
  }

  /** Create a contact (`409` if the number already exists). */
  create(params: CreateContactParams, options?: RequestOptions): Promise<Contact> {
    return this.one({ method: 'POST', path: '/v1/contacts', json: toWire(params), idempotent: true }, options);
  }

  get(id: string, options?: RequestOptions): Promise<Contact> {
    return this.one({ method: 'GET', path: `/v1/contacts/${seg(id)}` }, options);
  }

  /** Partial update: omitted fields are kept. */
  update(id: string, params: UpdateContactParams, options?: RequestOptions): Promise<Contact> {
    return this.one({ method: 'PATCH', path: `/v1/contacts/${seg(id)}`, json: toWire(params) }, options);
  }

  delete(id: string, options?: RequestOptions): Promise<void> {
    return this.none({ method: 'DELETE', path: `/v1/contacts/${seg(id)}` }, options);
  }
}

File: packages/rust/src/resources/contacts.rs
//! The `contacts` resource.

use super::{idem, list_query, seg};
use crate::error::Result;
use crate::http::{HttpClient, Idem};
use crate::models::{Contact, CreateContact, ListParams, Page, RequestOptions, UpdateContact};

/// Accessed as `client.contacts()`.
#[derive(Debug, Clone)]
pub struct Contacts {
    http: HttpClient,
}

impl Contacts {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// List contacts.
    pub async fn list(&self, params: ListParams) -> Result<Page<Contact>> {
        self.http
            .get(&format!("/v1/contacts{}", list_query(&params)))
            .await
    }

    /// Create a contact. A duplicate `e164` returns `409`.
    pub async fn create(&self, params: &CreateContact) -> Result<Contact> {
        self.create_with(params, &RequestOptions::default()).await
    }

    /// [`Contacts::create`] with an explicit Idempotency-Key.
    pub async fn create_with(
        &self,
        params: &CreateContact,
        opts: &RequestOptions,
    ) -> Result<Contact> {
        self.http
            .send("POST", "/v1/contacts", params, idem(opts))
            .await
    }

    /// Fetch a contact.
    pub async fn get(&self, id: &str) -> Result<Contact> {
        self.http
            .get(&format!("/v1/contacts/{}", seg(id, "id")?))
            .await
    }

    /// Update a contact (PATCH: unset fields are kept).
    pub async fn update(&self, id: &str, params: &UpdateContact) -> Result<Contact> {
        let path = format!("/v1/contacts/{}", seg(id, "id")?);
        self.http.send("PATCH", &path, params, Idem::None).await
    }

    /// Delete a contact.
    pub async fn delete(&self, id: &str) -> Result<()> {
        let path = format!("/v1/contacts/{}", seg(id, "id")?);
        self.http.empty("DELETE", &path, Idem::None).await
    }
}
